lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure/API/Address.hs

```diff
@@ -0,0 +1,183 @@
+    (hrp, dataPart) <- left InvalidBech32Encoding $ decodeLenient bech32
+    if hrp /= humanPart tag
+        then Left $ InvalidHumanReadablePart hrp
+        else case dataPartToBytes dataPart of
+            Nothing -> Left $ InvalidDataPart dataPart
+            Just bytes -> pure $ Shelley bytes
+
+decodeHumanAddress
+    :: NetworkTag -> Text -> Either DecodingError (AddressFlavor ByteString)
+decodeHumanAddress tag t =
+    decodeBech32Address tag t
+        <> decodeBase58Address (T.encodeUtf8 t)
+
+newtype CatchFail a = CatchFail {runCatchFail :: Either String a}
+    deriving (Functor, Applicative, Monad)
+
+instance MonadFail CatchFail where
+    fail = CatchFail . Left
+
+ledgerAddressFlavor :: SL.Addr c -> AddressFlavor ()
+ledgerAddressFlavor (SL.AddrBootstrap _) = Bootstrap ()
+ledgerAddressFlavor _ = Shelley ()
+
+ledgerAddressNetwworkTag :: SL.Addr c -> NetworkTag
+ledgerAddressNetwworkTag addr = case SL.getNetwork addr of
+    SL.Testnet -> TestnetTag
+    SL.Mainnet -> MainnetTag
+
+inspectAddress
+    :: NetworkTag
+    -> AddressFlavor ByteString
+    -> Either DecodingError (AddressFlavor Address)
+inspectAddress nt a = inspect $ addressFlavor a
+  where
+    inspect bs = do
+        r <-
+            left AddressDecodingError
+                $ runCatchFail
+                $ evalStateT
+                    (SH.decodeAddrStateLenientT @StandardCrypto True True bs)
+                    0
+        case (a, ledgerAddressNetwworkTag r, ledgerAddressFlavor r) of
+            (Bootstrap _, nt', Bootstrap ())
+                | nt == nt' -> pure $ Bootstrap $ SH.compactAddr r
+                | otherwise -> Left AddressNetworkMismatch
+            (Shelley _, nt', Shelley ())
+                | nt == nt' -> pure $ Shelley $ SH.compactAddr r
+                | otherwise -> Left AddressNetworkMismatch
+            _ -> Left AddressFlavorMismatch
+
+decodeFlavoredAddress
+    :: NetworkTag -> Text -> Either DecodingError (AddressFlavor Address)
+decodeFlavoredAddress nt = decodeHumanAddress nt >=> inspectAddress nt
+
+-- | Decode an 'Address' from a 'Text' representation given the 'NetworkTag'.
+decodeAddress :: NetworkTag -> Text -> Either DecodingError Address
+decodeAddress nt = fmap addressFlavor . decodeFlavoredAddress nt
+
+addFlavorToAddress :: Address -> AddressFlavor Address
+addFlavorToAddress x
+    | SL.isBootstrapCompactAddr x = Bootstrap x
+    | otherwise = Shelley x
+
+encodeFlavoredAddress :: NetworkTag -> AddressFlavor Address -> Text
+encodeFlavoredAddress tag (Shelley addr) = bech32
+  where
+    bytes = B8.fromShort $ toShortByteString addr
+    bech32 = Bech32.encodeLenient hrp (dataPartFromBytes bytes)
+    hrp = humanPart tag
+encodeFlavoredAddress _ (Bootstrap addr) =
+    T.decodeUtf8 . encodeBase58 bitcoinAlphabet
+        $ B8.fromShort
+        $ toShortByteString addr
+
+-- | Encode an 'Address' to a 'Text' representation given the 'NetworkTag'.
+encodeAddress :: NetworkTag -> Address -> Text
```

Comment:
you are big optimist here ... no chance to fail here? I would argue yes

Reply:
yes network can be wrong

Reply:
We removed the network constraints